daffodil-tdml-processor/src/main/scala/org/apache/daffodil/processor/tdml/DaffodilTDMLDFDLProcessor.scala

```diff
@@ -252,21 +250,47 @@ class DaffodilTDMLDFDLProcessor private (private var dp: DataProcessor)
     infosetXML: scala.xml.Node,
     outStream: java.io.OutputStream
   ): TDMLUnparseResult = {
-    val bos = new ByteArrayOutputStream()
-    val osw = new OutputStreamWriter(bos, StandardCharsets.UTF_8)
-    scala.xml.XML.write(osw, infosetXML, "UTF-8", xmlDecl = true, null)
-    osw.flush()
-    osw.close()
-    val saxInstream = new ByteArrayInputStream(bos.toByteArray)
-    doUnparseWithBothApis(inputter, saxInstream, outStream)
+    if (dp.tunables.testingParseUnparseAPIMode == Full) {
+      val bos = new ByteArrayOutputStream()
+      val osw = new OutputStreamWriter(bos, StandardCharsets.UTF_8)
+      scala.xml.XML.write(osw, infosetXML, "UTF-8", xmlDecl = true, null)
+      osw.flush()
+      osw.close()
+      val saxInstream = new ByteArrayInputStream(bos.toByteArray)
+      doUnparseWithBothApis(inputter, saxInstream, outStream)
+    } else {
+      doUnparseWithNonSaxApi(inputter, outStream)
+    }
+  }
+
+  def doParseWithNonSaxAPI(
```

Comment:
Feels like there's some duplicate code in these doParseWith*API functions since they both implement the non-sax API. Can they be combined into one function, e.g.

```scala
private def doParse(...) {
  // always do stuff with non-sax API

  if (apiMode = full) {
    // do sax API
    // verify SAX results match non-sax results
  }
}
```
Same with the unparse functions?